### PR TITLE
Fix reserved cgroup systemd

### LINKED ISSUE #78629: "--kube-reserved-cgroup" kubelet option cannot find cgroup
start kubelet with below config:
```
# root@k8s-node-prod-2:~# cat/etc/systemd/system/kubelet.service
Description=Kubernetes Kubelet
Documentation=https://github.com/GoogleCloudPlatform/kubernetes
After=docker.service
Requires=docker.service
[Service]
WorkingDirectory=/var/lib/kubelet
ExecStart=/usr/local/bin/kubelet \
  --address=10.x.x.x \
  --hostname-override=10.x.x.x \
  --cluster-dns=10.x.x.x \
  --experimental-bootstrap-kubeconfig=/etc/kubernetes/bootstrap.kubeconfig \
  --kubeconfig=/etc/kubernetes/kubelet.kubeconfig \
  --cert-dir=/etc/kubernetes/ssl \
  --cluster-domain=cluster.local. \
  --hairpin-mode promiscuous-bridge \
  --allow-privileged=true \
  --serialize-image-pulls=false \
  --logtostderr=true \
  --v=2 \
  --cgroup-driver=systemd \
  --cgroups-per-qos=true \
  --kube-reserved-cgroup=/podruntime.slice \
  --enforce-node-allocatable=pods,kube-reserved \
  --kube-reserved=cpu=500m,memory=500Mi,ephemeral-storage=10Gi 
  
Restart=on-failure
RestartSec=5
[Install]
WantedBy=multi-user.target
```
podruntime.slice  with below config:
```
# root@k8s-node-prod-2:~# cat /lib/systemd/system/podruntime.slice
[Unit]
Description=Limited resources slice for Kubernetes services
Documentation=man:systemd.special(7)
DefaultDependencies=no
Before=slices.target
Requires=-.slice
After=-.slice
```
check the podruntime.slice  cgroup, get the following output:
```
#root@k8s-node-prod-2:~# systemctl status podruntime.slice
● podruntime.slice - Limited resources slice for Kubernetes services
   Loaded: loaded (/lib/systemd/system/podruntime.slice; static; vendor preset: enabled)
   Active: active since Sun 2019-06-02 11:29:57 CST; 42min ago
     Docs: man:systemd.special(7)
    Tasks: 0
   Memory: 0B
      CPU: 0
   CGroup: /podruntime.slice

Jun 02 11:29:57 k8s-node-prod-2 systemd[1]: Created slice Limited resources slice for Kubernetes services.
```
then, start kubelet.service, run `journalctl -u kubelet.service  -f`  as follows:
```
Jun 02 12:36:40 k8s-node-prod-2 kubelet[73505]: I0602 12:36:40.187096   73505 node_container_manager_linux.go:114] Enforcing kube reserved on cgroup "/podruntime.slice" with limits: map[cpu:{i:{value:500 scale:-3} d:{Dec:<nil>} s:500m Format:DecimalSI} ephemeral-storage:{i:{value:10737418240 scale:0} d:{Dec:<nil>} s:10Gi Format:BinarySI} memory:{i:{value:524288000 scale:0} d:{Dec:<nil>} s:500Mi Format:BinarySI}]
Jun 02 12:36:40 k8s-node-prod-2 kubelet[73505]: F0602 12:36:40.187428   73505 kubelet.go:1359] Failed to start ContainerManager Failed to enforce Kube Reserved Cgroup Limits on "/podruntime.slice": ["podruntime.slice"] cgroup does not exist
Jun 02 12:36:40 k8s-node-prod-2 systemd[1]: kubelet.service: Main process exited, code=exited, status=255/n/a
Jun 02 12:36:40 k8s-node-prod-2 systemd[1]: kubelet.service: Unit entered failed state.
Jun 02 12:36:40 k8s-node-prod-2 systemd[1]: kubelet.service: Failed with result 'exit-code'.
Jun 02 12:36:45 k8s-node-prod-2 systemd[1]: kubelet.service: Service hold-off time over, scheduling restart.
```
`["podruntime.slice"] cgroup does not exist`
why?

**Environment**:
- Kubernetes version (use `kubectl version`): v1.14.1
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Debian GNU/Linux 9 (stretch)
- Kernel (e.g. `uname -a`):  SMP Debian 4.9.30-2+deb9u2 (2017-06-26) x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by mattjmcnaughton**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fix an issue in which, when trying to specify the `--kube-reserved-cgroup`
(or `--system-reserved-cgroup`) with `--cgroup-driver=systemd`, we will
not properly convert the `systemd` cgroup name into the internal cgroup
name that k8s expects. Without this change, specifying
`--kube-reserved-cgroup=/test.slice --cgroup-driver=systemd` will fail,
and only `--kube-reserved-cgroup=/test --crgroup-driver=systemd` will succeed,
even if the actual cgroup existing on the host is `/test.slice`.

Additionally, add light unit testing of our process from converting to a
systemd cgroup name to kubernetes internal cgroup name.

**Which issue(s) this PR fixes**:
Fixes #78629 

**Special notes for your reviewer**:
cc @xiongmaodada @jaypipes 

**Does this PR introduce a user-facing change?**:
```release-note
When specifying `--(kube|system)-reserved-cgroup`, with `--cgroup-driver=systemd`, it is now possible to use the fully qualified cgroupfs name (i.e. `/test-cgroup.slice`).
```

### CODE REVIEW

**Comment by mattjmcnaughton** [comment]
/assign @ConnorDoyle 

thanks :)

**Comment by dashpole** [comment]
While it might not have been the optimal way to implement it, I believe the intended way to set the flag is to specify `--kube-reserved-cgroup=/test` to use the `/test.slice` cgroup.

cc @derekwaynecarr who might remember why we did this, or if it was a mistake we just decided to stick with for backwards compatibility.

**Comment by dashpole** [comment]
I took a look at this PR, and I don't think this meets backwards compatibility requirements for kubelet flags.  For example, if someone is currently specifying `--kube-reserved-cgroup=/test`, their kubelet wouldn't work after upgrading unless they switched to  `--kube-reserved-cgroup=/test.slice`.

**Comment by jaypipes** [comment]
Hmm, I was thinking a good solution to the original bug (from a user experience perspective) would be to simply have the configuration option handler for `--kube-reserved-group` simply strip off any trailing ".slice" notation that the user may have erroneously supplied.

That would be backwards-compatible and give a better user experience than the error message that was returned in #78629

**Comment by mattjmcnaughton** [comment]
@ConnorDoyle @jaypipes thanks for the feedback!

The `CgroupName` function calls `ParseSystemdToCgroupName`, which strips the `.slice` only if it exists. So I think this pr may already be doing what @jaypipes is describing. In addition, I believe this diff will not harm backwards compatibility, because both `/test` and `/test.slice` can be passed.

Very possible I'm misreading the code though :)

**Comment by dashpole** [comment]
Ah, you are correct.  In that case, this change
/lgtm

**Review by jaypipes — APPROVED** [review]
Suffice to say it's a bit confusing to have a method called `CgroupName()` on the cgroupManager object in addition to a type named `CgroupName` :)

But yes, this patch works. Might be useful to mention in the commit message that the `CgroupName` method of the cgroupManager object calls `ParseSystemdToCgroupName()`, which is the part that fixes the bug/issue in question.

**Comment by mattjmcnaughton** [comment]
Thanks all :)

cc @ConnorDoyle friendly ping :)

**Comment by mattjmcnaughton** [comment]
cc @Random-Liu mind taking a look and approving if you get the chance?

**Comment by mattjmcnaughton** [comment]
/assign @vishh 

mind taking a look if you get the chance? would love to get this bug fixed and the pr closed :)

**Comment by mattjmcnaughton** [comment]
cc @tallclair mind approving? would love to get this merged and the issue closed.

**Comment by mattjmcnaughton** [comment]
/assign @derekwaynecarr 

Could you please take a look? Thanks! I want to rule this out as a cause of #80807.

**Comment by derekwaynecarr** [comment]
I will take a look.

@dashpole I need to verify the behavior intended.  Internally, we had the kubelet describe cgroups abstractly into the cgroupfs format, and converted it to a concrete name.  Externally, I had thought we settled on using cgroupfs name, but yeah, this will be a fun walk down memory lane to remember what was intended or not ;-)

**Comment by derekwaynecarr** [comment]
@mattjmcnaughton its also possible global accounting was not enabled on the CoreOS host.

see:
https://www.freedesktop.org/software/systemd/man/systemd-system.conf.html#DefaultCPUAccounting=

**Comment by derekwaynecarr** [comment]
The original design had stated that only the cgroupfs fully qualified name was to be used.

See:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-allocatable.md#phase-2---enforce-allocatable-on-pods

**Comment by derekwaynecarr** [comment]
/hold cancel

This fixes a regression I suspect that was introduced when we converted how we managed internal cgroup names.  I do not think it will address the CoreOS issue that was noted as I think that is due to not having DefaultAccounting enabled on systemd for the unit that launched the kubelet.

Either way, this is a net win.

/approve
/lgtm

**Comment by derekwaynecarr** [comment]
Updated the release note for clarity that the user is intended to use the fully qualified cgroupfs name.

**Comment by derekwaynecarr** [comment]
/hold 

actually, this confused me for a second.

**Comment by derekwaynecarr** [comment]
/hold cancel

the desired behavior was to always specify the fully qualified name and never the internal name, so always /system.slice and never /system even when using systemd cgroup driver.  i think this regression was introduced when @filbranden tried to clean up our internal cgroup naming logic.  In practice, it is extremely rare and not really recommended to enforce reservations on system or kube-reserved groups (which is why its not the default), which explains why this was probably not found previously.  Its actually a good idea for us to log a warning if people do enforce it that tehy really really konw what they are doing!

**Comment by CamelCaseNotation** [comment]
> /hold cancel
> 
> the desired behavior was to always specify the fully qualified name and never the internal name, so always /system.slice and never /system even when using systemd cgroup driver. i think this regression was introduced when @filbranden tried to clean up our internal cgroup naming logic. In practice, it is extremely rare and not really recommended to enforce reservations on system or kube-reserved groups (which is why its not the default), which explains why this was probably not found previously. Its actually a good idea for us to log a warning if people do enforce it that tehy really really konw what they are doing!

Can you elaborate? Why is it not recommended to do that?

**Pull request merged**
